pkg/kube/multicluster/secretcontroller.go

```diff
@@ -444,6 +449,66 @@ var BuildClientsFromConfig = func(kubeConfig []byte) (kube.Client, error) {
 	return clients, nil
 }
 
+// sanitizeKubeConfig sanitizes a kubeconfig file to strip out insecure settings which may leak
+// confidential materials.
+// See https://github.com/kubernetes/kubectl/issues/697
+func sanitizeKubeConfig(config api.Config, allowlist sets.Set) error {
+	for k, auths := range config.AuthInfos {
+		if ap := auths.AuthProvider; ap != nil {
+			// We currently are importing 5 authenticators: gcp, azure, exec, and openstack
+			switch ap.Name {
+			case "oidc":
+				// OIDC is safe as it doesn't read files or execute code.
+				// create-remote-secret specifically supports OIDC so its probably important to not break this.
+			default:
+				if !allowlist.Contains(ap.Name) {
+					// All the others - gcp, azure, exec, and openstack - are unsafe
+					return fmt.Errorf("auth provider %s is not allowed", ap.Name)
+				}
+			}
+		}
+		if auths.ClientKey != "" && !allowlist.Contains("clientKey") {
+			return fmt.Errorf("clientKey is not allowed")
+		}
+		if auths.ClientCertificate != "" && !allowlist.Contains("clientCertificate") {
+			return fmt.Errorf("clientCertificate is not allowed")
+		}
```

Comment:
Hey @howardjohn,

Can you provide more information about how having this (paths to a tls client cert and key) is insecure?

I looked at https://github.com/kubernetes/kubectl/issues/697 but couldn't see anything related to how these can be leaked.

Reply:
Attack vector is a user creates a kubeconfig with `clientKey: /some/path/in/istiod/container`. Then they set the API server URL to some server that records the key sent. This allows ex-filtrating arbitrary files from the Istio filesystem.

Reply:
`clientKey` is the path to the TLS private key, it's not sent over the wire, correct?

---

The argument may apply to `clientCertificate` which does get sent over the wire. Though even there I think the TLS stack will not sent arbitrary bytes over the wire and we will fail validation (e.g. [in here](https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/crypto/tls/tls.go;l=265))

Reply:
Correct, I should have said clientCertificate - it may not allow _arbitrary_ exfiltration in this case, but is still a SSRF. For example, Istiod may have high privileged mtls certs, and now a user can have them send requests using those credentials to arbitrary services.

If you have legitimate use cases you should be able to re-enable this, right?

Reply:
Yeah, I can re-enable this. I just wanted to understand what I am re-enabling. Thanks for the explanation.

Reply:
:+1: . Token auth is much worse than mTLS